File: ebanking-backend/src/main/java/ma/enset/ebankingbackend/services/BankAccountService.java
package ma.enset.ebankingbackend.services;

import ma.enset.ebankingbackend.dtos.*;
import ma.enset.ebankingbackend.entities.BankAccount;
import ma.enset.ebankingbackend.entities.CurrentAccount;
import ma.enset.ebankingbackend.entities.Customer;
import ma.enset.ebankingbackend.entities.SavingAccount;
import ma.enset.ebankingbackend.exceptions.BalanceNotSufficientException;
import ma.enset.ebankingbackend.exceptions.BankAccountNotFoundException;
import ma.enset.ebankingbackend.exceptions.CustomerNotFoundException;

import java.util.List;

public interface BankAccountService {
    CustomerDTO saveCustomer(CustomerDTO customerDTO);

    CustomerDTO updateCustomer(CustomerDTO customerDTO);

    void deleteCustomer(Long customerId);

    CurrentBankAccountDTO saveCurrentBankAccount(double initialBalance, double overDraft, Long customerId) throws CustomerNotFoundException;
    SavingBankAccountDTO saveSavingBankAccount(double initialBalance, double interestRate, Long customerId) throws CustomerNotFoundException;

    List<CustomerDTO> listCustomers();
    BankAccountDTO getBankAccount(String accountId) throws BankAccountNotFoundException;

    BankAccountDTO updateBankAccount(BankAccountDTO bankAccountDTO) throws BankAccountNotFoundException;

    void deleteBankAccount(String accountId);

    void debit(String accountId, double amount, String description) throws BankAccountNotFoundException, BalanceNotSufficientException;
    void credit(String accountId, double amount,String description) throws BankAccountNotFoundException;
    void transfer(String accountIdSource, String accountIdDestination, Double amount) throws BankAccountNotFoundException, BalanceNotSufficientException;

    List<BankAccountDTO> bankAccountList();
    List<BankAccountDTO> bankAccountList(Long customerId);
    CustomerDTO getCustomer(Long customerId) throws CustomerNotFoundException;

    List<AccountOperationDTO> accountHistorique(String accountId);

    AccountHistoryDTO getAccountHistory(String accountId, int page, int size) throws BankAccountNotFoundException;

    List<CustomerDTO> searchCustomers(String keyword);

    CustomerDTO searchCustomerById(Long id) throws CustomerNotFoundException;
}

File: ebanking-frontend/src/app/services/bank-account.service.ts
import { inject, Injectable } from '@angular/core';
import { environment } from '../environments/environment';
import { HttpClient } from '@angular/common/http';
import { BankAccount } from '../models/bank-account.model';
import { Observable } from 'rxjs';
import { SavingAccountRequestDTO } from '../dtos/SavingAccountRequestDTO';
import { CurrentAccountRequestDTO } from '../dtos/CurrentAccountRequestDTO';
import { BankAccountDTO } from '../dtos/BankAccountDTO';
import { AccountHistoryDTO } from '../dtos/AccountHistoryDTO';

@Injectable({
  providedIn: 'root'
})
export class BankAccountService {

  backendUrl = environment.baseUrl + '/accounts';
  private http = inject(HttpClient);

  getBankAccounts(): Observable<BankAccount[]> {
    return this.http.get<BankAccount[]>(`${this.backendUrl}`);
  }

  getBankAccountById(id: string): Observable<BankAccountDTO> {
    return this.http.get<BankAccountDTO>(`${this.backendUrl}/${id}`);
  }

  getBankAccountsByCustomerId(customerId: number): Observable<BankAccount[]> {
    return this.http.get<BankAccount[]>(`${this.backendUrl}/by-customer/${customerId}`);
  }

  saveSavingBankAccount(savingBankAccountDTO: SavingAccountRequestDTO): Observable<SavingAccountRequestDTO> {
    return this.http.post<SavingAccountRequestDTO>(`${this.backendUrl}/saving`, savingBankAccountDTO);
  }

  saveCurrentBankAccount(currentBankAccountDTO: CurrentAccountRequestDTO): Observable<CurrentAccountRequestDTO> {
    return this.http.post<CurrentAccountRequestDTO>(`${this.backendUrl}/current`, currentBankAccountDTO);
  }

  updateBankAccount(bankAccountDTO: BankAccountDTO): Observable<BankAccountDTO> {
    return this.http.put<BankAccountDTO>(`${this.backendUrl}/${bankAccountDTO.id}`, bankAccountDTO);
  }

  deleteBankAccount(id: string): Observable<void> {
    return this.http.delete<void>(`${this.backendUrl}/${id}`);
  }


  getAccountHistory(accountId: string, page: number, size: number): Observable<AccountHistoryDTO> {
    return this.http.get<AccountHistoryDTO>(
      `${this.backendUrl}/${accountId}/pageOperations?page=${page}&size=${size}`
    );
  }

  debit(accountId: string, amount: number, description: string): Observable<void> {
    return this.http.post<void>(`${this.backendUrl}/debit`, {
      accountId, amount, description
    });
  }

  transfer(sourceId: string, destId: string, amount: number): Observable<void> {
    return this.http.post<void>(`${this.backendUrl}/transfer`, {
      sourceAccount: sourceId,
      destinationAccount: destId,
      amount
    });
  }


  getAllAccountsExcept(currentId: string): Observable<BankAccountDTO[]> {
    return this.http.get<BankAccountDTO[]>(`${this.backendUrl}/others/${currentId}`);
  }

  credit(accountId: string, amount: number, description: string): Observable<void> {
    return this.http.post<void>(`${this.backendUrl}/credit`, {
      accountId, amount, description
    });
  }

}
